plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs

```diff
@@ -0,0 +1,225 @@
+           -> Type TyName DefaultUni ()
+substType' nested sub ty0 = go fvs Set.empty sub ty0
+  where
+    fvs = Set.unions $ Map.keysSet sub : map (setOf ftvTy) (Map.elems sub)
+
+    go :: HasCallStack => _
+    go fvs seen sub ty = case ty of
+      TyVar _ x | Set.member x seen -> error "substType' loop"
+      -- In the case where we do nested substitution we just continue, in parallel substitution
+      -- we never go below a substitution.
+      TyVar _ x | nested    -> maybe ty (go fvs (Set.insert x seen) sub) $ Map.lookup x sub
+                | otherwise -> maybe ty id $ Map.lookup x sub
+      TyFun _ a b      -> TyFun () (go fvs seen sub a) (go fvs seen sub b)
+      TyApp _ a b      -> TyApp () (go fvs seen sub a) (go fvs seen sub b)
+      TyLam _ x k b
+        | Set.member x fvs -> TyLam () x' k $ go (Set.insert x' fvs) seen sub (renameType x x' b)
+        | otherwise        -> TyLam () x  k $ go (Set.insert x fvs) (Set.delete x seen) sub b
+        where x' = freshenTyName (fvs <> setOf ftvTy b) x
+      TyForall _ x k b
+        | Set.member x fvs -> TyForall () x' k $ go (Set.insert x' fvs) seen sub (renameType x x' b)
+        | otherwise        -> TyForall () x  k $ go (Set.insert x fvs) (Set.delete x seen) sub b
+        where x' = freshenTyName (fvs <> setOf ftvTy b) x
+      TyBuiltin{}      -> ty
+      TyIFix _ a b     -> TyIFix () (go fvs seen sub a) (go fvs seen sub b)
+
+-- | Find all free type variables of type `a` given substitution `sub`. If variable `x` is
+-- free in `a` but in the domain of `sub` we look up `x` in `sub` and get all the free type
+-- variables of the result - up to the substitution.
+fvTypeR :: Map TyName (Type TyName DefaultUni ()) -> Type TyName DefaultUni () -> Set TyName
+fvTypeR sub a = Set.unions $ freeAndNotInSub : map (fvTypeR sub . (Map.!) sub) (Set.toList freeButInSub)
+      where
+          fvs = setOf ftvTy a
+          subDom = Map.keysSet sub
+          freeButInSub = Set.intersection subDom fvs
+          freeAndNotInSub = Set.difference fvs subDom
+
+-- * Generators for substitutions
+
+-- | Generate a type substitution that is valid in a given context.
+genSubst :: Map TyName (Kind ()) -> Gen (Map TyName (Type TyName DefaultUni ()))
+genSubst ctx = do
+  xks <- sublistOf <=< shuffle $ Map.toList ctx
+  go ctx Map.empty xks
+  where
+    -- Counts is used to balance the ratio between the number
+    -- of times a variable `x` appears in the substitution and
+    -- the size of the type it maps to - the more times `x` appears
+    -- the smaller the type it maps to needs to be to avoid blowup.
+    go _ _ [] = return mempty
+    go ctx counts ((x, k) : xs) = do
+      let ctx' = Map.delete x ctx
+          w    = fromMaybe 1 $ Map.lookup x counts
+      ty <- sized $ \ n -> resize (div n w) $ genTypeWithCtx ctx' k
+      let moreCounts = fmap (* w) $ fvTypeBag ty
+          counts'    = Map.unionWith (+) counts moreCounts
+      Map.insert x ty <$> go ctx' counts' xs
```

Comment:
Note to self: explain this in a bit more words.

Reply:
> Note to self: explain this in a bit more words.

[Done](https://github.com/input-output-hk/plutus/pull/4559/commits/4ef3ca6db831027eafb9507f93c02ca20db53f85). And also made the function tail-recursive while documenting it, because referring to types that are not yet inserted into the substitution but already affect the logic was awkward.